File: python/src/sci_task_io/ssts/validation.py
"""Validation helpers for SSTS and processed signal NPY payloads."""

import re

from .ssts import SSTS, SignalTrack

_TRACK_KEY_RE = re.compile(r"^track_([1-9][0-9]*)$")


class ValidationError(ValueError):
    """
    Raised when payload validation fails.

    Notes:
    - This is the canonical validation exception for schema and runtime checks.
    """


def _is_scalar(value):
    """
    Check whether a value is a scalar JSON-compatible value.

    Behavior:
    - Accepts: number, string, boolean, null.
    - Returns True when accepted by `scalars` field contract.
    """
    return value is None or isinstance(value, (bool, int, float, str))


def _is_time_value(value):
    """
    Check whether a value is a valid time value.

    Behavior:
    - Accepts: number or string.
    - Used to validate `times` arrays.
    """
    return isinstance(value, (int, float, str))


def _validate_monotonic(times, require_ordering):
    """
    Validate monotonic non-decreasing time ordering when requested.

    Parameters:
    - `times`: sequence of time values.
    - `require_ordering`: if False, ordering checks are skipped.
    - Raises ValidationError on decreasing or non-comparable adjacent values.
    """
    if not require_ordering or len(times) < 2:
        return
    for left, right in zip(times[:-1], times[1:]):
        try:
            if left > right:
                raise ValidationError("times must be monotonic non-decreasing.")
        except TypeError as exc:
            raise ValidationError("times contains non-comparable values.") from exc


def validate_signal_track_payload(payload, require_ordering=False):
    """
    Validate one `track_n` payload.

    Behavior:
    - Requires keys: `label`, `times`, `signal`.
    - Enforces non-empty string label.
    - Enforces `times` and `signal` arrays with equal length.
    - Optionally enforces monotonic ordering of `times`.
    """
    if not isinstance(payload, dict):
        raise ValidationError("track payload must be an object.")

    if "label" not in payload or "times" not in payload or "signal" not in payload:
        raise ValidationError("track payload must contain label, times, and signal.")

    label = payload["label"]
    times = payload["times"]
    signal = payload["signal"]

    if not isinstance(label, str) or not label:
        raise ValidationError("track label must be a non-empty string.")
    if not isinstance(times, list):
        raise ValidationError("track times must be an array.")
    if not isinstance(signal, list):
        raise ValidationError("track signal must be an array.")
    if len(times) != len(signal):
        raise ValidationError("track times and signal lengths must match.")

    for value in times:
        if not _is_time_value(value):
            raise ValidationError("time values must be numbers or strings.")

    _validate_monotonic(times, require_ordering=require_ordering)


def validate_ssts_payload(payload, require_ordering=False):
    """
    Validate full SSTS payload for `<series_id>.json`.

    Behavior:
    - Requires top-level `signals` object.
    - Validates optional `metadata` and `scalars` object types.
    - Enforces `signals` keys match `track_<positive_integer>`.
    - Enforces contiguous track numbering `track_1..track_N`.
    - Validates each track via `validate_signal_track_payload`.
    """
    if not isinstance(payload, dict):
        raise ValidationError("SSTS payload must be an object.")

    if "signals" not in payload:
        raise ValidationError("SSTS payload must contain signals.")

    metadata = payload.get("metadata", {})
    scalars = payload.get("scalars", {})
    signals = payload["signals"]

    if metadata is not None and not isinstance(metadata, dict):
        raise ValidationError("metadata must be an object when provided.")
    if scalars is not None and not isinstance(scalars, dict):
        raise ValidationError("scalars must be an object when provided.")

    if isinstance(scalars, dict):
        for key, value in scalars.items():
            if not isinstance(key, str):
                raise ValidationError("scalars keys must be strings.")
            if not _is_scalar(value):
                raise ValidationError("scalars values must be scalar JSON values.")

    if not isinstance(signals, dict):
        raise ValidationError("signals must be an object.")

    track_numbers = []
    for key, track_payload in signals.items():
        match = _TRACK_KEY_RE.match(key)
        if not match:
            raise ValidationError(f"invalid track key: {key!r}.")
        track_numbers.append(int(match.group(1)))
        validate_signal_track_payload(track_payload, require_ordering=require_ordering)

    if track_numbers:
        expected = list(range(1, max(track_numbers) + 1))
        if sorted(track_numbers) != expected:
            raise ValidationError("track keys must be contiguous from track_1 to track_N.")


def ssts_from_payload(payload, require_ordering=False):
    """
    Build an `SSTS` object from raw payload after validation.

    Behavior:
    - Validates payload first.
    - Preserves unknown top-level and track-level fields in `.extra`.
    - Returns normalized in-memory `SSTS` and `SignalTrack` objects.
    """
    validate_ssts_payload(payload, require_ordering=require_ordering)

    known = {"metadata", "scalars", "signals"}
    extra = {k: v for k, v in payload.items() if k not in known}

    signals = payload["signals"]
    tracks = []
    for idx in range(1, len(signals) + 1):
        track_obj = signals[f"track_{idx}"]
        track_extra = {k: v for k, v in track_obj.items() if k not in {"label", "times", "signal"}}
        tracks.append(
            SignalTrack(
                label=track_obj["label"],
                times=list(track_obj["times"]),
                signal=list(track_obj["signal"]),
                extra=track_extra,
            )
        )

    metadata = payload.get("metadata") or {}
    scalars = payload.get("scalars") or {}
    return SSTS(tracks=tracks, metadata=dict(metadata), scalars=dict(scalars), extra=extra)


def validate_signal_label_npy_payload(payload, filename_stem=None, require_ordering=False):
    """
    Validate one processed `<signal_label>.npy` payload.

    Behavior:
    - Applies standard track validation (`label`, `times`, `signal`).
    - If `filename_stem` is provided, enforces `payload['label'] == filename_stem`.
    """
    validate_signal_track_payload(payload, require_ordering=require_ordering)
    if filename_stem is not None and payload["label"] != filename_stem:
        raise ValidationError("npy payload label must match filename stem.")

File: rust/src/ssts/validation.rs
use std::cmp::Ordering;
use std::collections::BTreeSet;
use std::error::Error;
use std::fmt::{Display, Formatter};

use serde_json::{Map, Value};

use super::ssts::{SSTS, SignalTrack};

#[derive(Debug, Clone)]
pub struct ValidationError {
    pub message: String,
}

impl ValidationError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl Display for ValidationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

impl Error for ValidationError {}

fn compare_time_values(left: &Value, right: &Value) -> Result<Ordering, ValidationError> {
    match (left, right) {
        (Value::Number(a), Value::Number(b)) => {
            let a = a
                .as_f64()
                .ok_or_else(|| ValidationError::new("invalid numeric time value"))?;
            let b = b
                .as_f64()
                .ok_or_else(|| ValidationError::new("invalid numeric time value"))?;
            Ok(a.partial_cmp(&b).unwrap_or(Ordering::Equal))
        }
        (Value::String(a), Value::String(b)) => Ok(a.cmp(b)),
        (Value::Number(_), Value::String(_)) => Ok(Ordering::Less),
        (Value::String(_), Value::Number(_)) => Ok(Ordering::Greater),
        _ => Err(ValidationError::new("time values must be numbers or strings")),
    }
}

fn validate_time_values(times: &[Value], require_ordering: bool) -> Result<(), ValidationError> {
    for value in times {
        if !value.is_number() && !value.is_string() {
            return Err(ValidationError::new("time values must be numbers or strings"));
        }
    }

    if require_ordering {
        for pair in times.windows(2) {
            if compare_time_values(&pair[0], &pair[1])? == Ordering::Greater {
                return Err(ValidationError::new(
                    "times must be monotonic non-decreasing",
                ));
            }
        }
    }

    Ok(())
}

fn parse_track_key(key: &str) -> Option<usize> {
    let suffix = key.strip_prefix("track_")?;
    if suffix.is_empty() {
        return None;
    }
    let n = suffix.parse::<usize>().ok()?;
    if n == 0 {
        return None;
    }
    Some(n)
}

fn validate_signal_track_payload(
    payload: &Map<String, Value>,
    require_ordering: bool,
) -> Result<(), ValidationError> {
    let label = payload
        .get("label")
        .ok_or_else(|| ValidationError::new("track payload must contain label"))?;
    let times = payload
        .get("times")
        .ok_or_else(|| ValidationError::new("track payload must contain times"))?;
    let signal = payload
        .get("signal")
        .ok_or_else(|| ValidationError::new("track payload must contain signal"))?;

    let label = label
        .as_str()
        .ok_or_else(|| ValidationError::new("track label must be a non-empty string"))?;
    if label.is_empty() {
        return Err(ValidationError::new("track label must be a non-empty string"));
    }

    let times = times
        .as_array()
        .ok_or_else(|| ValidationError::new("track times must be an array"))?;
    let signal = signal
        .as_array()
        .ok_or_else(|| ValidationError::new("track signal must be an array"))?;

    if times.len() != signal.len() {
        return Err(ValidationError::new(
            "track times and signal lengths must match",
        ));
    }

    validate_time_values(times, require_ordering)
}

fn validate_scalars(scalars: &Map<String, Value>) -> Result<(), ValidationError> {
    for (k, v) in scalars {
        if k.is_empty() {
            return Err(ValidationError::new("scalars keys must be non-empty strings"));
        }
        if !(v.is_number() || v.is_string() || v.is_boolean() || v.is_null()) {
            return Err(ValidationError::new(
                "scalars values must be scalar JSON values",
            ));
        }
    }
    Ok(())
}

pub fn validate_ssts_payload(payload: &Value, require_ordering: bool) -> Result<(), ValidationError> {
    let obj = payload
        .as_object()
        .ok_or_else(|| ValidationError::new("SSTS payload must be an object"))?;

    if let Some(metadata) = obj.get("metadata") {
        if !metadata.is_object() {
            return Err(ValidationError::new("metadata must be an object when provided"));
        }
    }

    if let Some(scalars) = obj.get("scalars") {
        let scalars = scalars
            .as_object()
            .ok_or_else(|| ValidationError::new("scalars must be an object when provided"))?;
        validate_scalars(scalars)?;
    }

    let signals = obj
        .get("signals")
        .ok_or_else(|| ValidationError::new("SSTS payload must contain signals"))?
        .as_object()
        .ok_or_else(|| ValidationError::new("signals must be an object"))?;

    let mut ids = BTreeSet::new();
    for (key, track) in signals {
        let id = parse_track_key(key)
            .ok_or_else(|| ValidationError::new(format!("invalid track key: {key:?}")))?;
        ids.insert(id);

        let track = track
            .as_object()
            .ok_or_else(|| ValidationError::new("track payload must be an object"))?;
        validate_signal_track_payload(track, require_ordering)?;
    }

    if let Some(max_id) = ids.iter().max().copied() {
        for expected in 1..=max_id {
            if !ids.contains(&expected) {
                return Err(ValidationError::new(
                    "track keys must be contiguous from track_1 to track_N",
                ));
            }
        }
    }

    Ok(())
}

pub fn ssts_from_payload(payload: &Value, require_ordering: bool) -> Result<SSTS, ValidationError> {
    validate_ssts_payload(payload, require_ordering)?;

    let obj = payload
        .as_object()
        .ok_or_else(|| ValidationError::new("SSTS payload must be an object"))?;

    let metadata = obj
        .get("metadata")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();
    let scalars = obj
        .get("scalars")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_default();

    let known = ["metadata", "scalars", "signals"];
    let mut extra = Map::new();
    for (k, v) in obj {
        if !known.contains(&k.as_str()) {
            extra.insert(k.clone(), v.clone());
        }
    }

    let signals = obj
        .get("signals")
        .and_then(Value::as_object)
        .ok_or_else(|| ValidationError::new("signals must be an object"))?;

    let mut tracks = Vec::new();
    for idx in 1..=signals.len() {
        let key = format!("track_{idx}");
        let track_obj = signals
            .get(&key)
            .and_then(Value::as_object)
            .ok_or_else(|| ValidationError::new(format!("missing contiguous track: {key}")))?;

        let label = track_obj
            .get("label")
            .and_then(Value::as_str)
            .ok_or_else(|| ValidationError::new("track label must be a non-empty string"))?
            .to_string();
        let times = track_obj
            .get("times")
            .and_then(Value::as_array)
            .ok_or_else(|| ValidationError::new("track times must be an array"))?
            .clone();
        let signal = track_obj
            .get("signal")
            .and_then(Value::as_array)
            .ok_or_else(|| ValidationError::new("track signal must be an array"))?
            .clone();

        let mut track_extra = Map::new();
        for (k, v) in track_obj {
            if k != "label" && k != "times" && k != "signal" {
                track_extra.insert(k.clone(), v.clone());
            }
        }

        tracks.push(SignalTrack {
            label,
            times,
            signal,
            extra: track_extra,
        });
    }

    Ok(SSTS {
        tracks,
        metadata,
        scalars,
        extra,
    })
}
